### ISSUE TITLE
use view blocks generated by cell within main layout/vew/request

### ISSUE THREAD

**Issue by saeideng** [cakephp/cakephp#10166]
This is a (multiple allowed):

* [ ] bug
* [x] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.4.x
### What you did
A way to render blocks(like js making with scriptBlock() ) within main layout
```
//cell
$this->Html->scriptBlock('alert("hi")', ['block' => 'blockName']);
//layout : outside of cell
$this->fetch('blockName');
```
I can explain more with sample if you need

**Comment by markstory** [comment]
I don't really understand what you're trying to do here.

**Comment by dereuromark** [comment]
I think he wants to share view blocks between cells and the main view rendering process.

**Comment by saeideng** [comment]
> I think he wants to share view blocks between cells and the main view rendering process.

yes

**Comment by dereuromark** [comment]
I do not think the cells are designed for this, they should be small independent sub-renderings so to speak.
If you need to share things, then maybe it should all be in the main one, using helpers, elements and alike instead.

**Comment by thinkingmedia** [comment]
You have one cell render the html for the cell, and a second call render the Javascript for the cell.

**Comment by saeideng** [comment]
> You have one call render the html for the cell, and a second call render the Javascript for the cell.

if js vars/Html created by model data( DB/... ) we need access 2 time to DB(for example)
I know we can using cache But ... is not fast way for developing :smiley:

**Comment by thinkingmedia** [comment]
I'm sorry, it wasn't the best solution but that's how you have to do it with the current way it is. I have a graph cell that has to embed the script inline because of this limitation. So I feel your pain.

We need a solution that's compatible with caching.

How about a new view class that you can use with a cell that returns an object instead of a string, and that object has a `getBody()` and `getBlock($name)` methods. This object could be serializable to continue support for caching.

**Comment by markstory** [comment]
> I do not think the cells are designed for this, they should be small independent sub-renderings so to speak.

They are not cells are intentionally isolated from the main rendering flow. If you want to share view context then use elements.

**Comment by thinkingmedia** [comment]
@markstory yes an element solves the view context, but then you loose the ability to perform queries on a model. Think of a Graph cell as an example. You have to render the HTML for the canvas, but you also want to append the JavaScript code at the bottom of the layout, and you have to query the data for the graph. The graph itself is heavy so you want to cache this.

Element won't solve cache because of the scripts block, and a cell can't use blocks.

What's the solution?

**Comment by dereuromark** [comment]
You could probably hack it by sharing it via Configure and alike. Anything that is actually shared inside the same dispatcher request.
But I do not think that is a very good framework approach here.

**Comment by thinkingmedia** [comment]
I'm going to give a crack at a view class to solve this particular problem, and reviewing a PR might be easier for discussion (in this case).

**Comment by thinkingmedia** [comment]
Oh I figured it out! You can do it without any changes. We just need to document this better.

An example `index.ctp` using a cell:

```
   $cached = Cache::remember('myCell', function() {
        $view = $this->cell('Example')->createView();
        return [
            'html' => $view->render(),
            'script' => $view->Blocks->get('script')
        ];
   });

  echo $cached['html'];
  $this->Blocks->concat('script', $cached['script']);
```

I think the above would solve the problem. We need some way of making this easy for people to do. 

I was thinking we could add a 'cache' parameter to the options array when calling `View::cell` but the problem there is that the cell method returns an object.

I'll think on this for a bit.

**Comment by inoas** [comment]
Right now with the tools out of the box you could not create a script block that you run at say scriptBottom from within a cell view template, right?

> >    I do not think the cells are designed for this, they should be small independent sub-renderings so to speak.

> They are not cells are intentionally isolated from the main rendering flow. If you want to share view context then use elements.

Maybe it would be good if there was something like 'viewScope' attribute for those HtmlHelper methods that allow to append css code/ressources above/below the fold?

**Comment by markstory** [comment]
You could also pass the current view into the cell [as an argument](https://book.cakephp.org/3.0/en/views/cells.html#passing-arguments-to-a-cell). You'd then be able to use the 'parent' view's helpers. Something like:

```php
class SharedStateCell
{
  public function share($parentView)
  {
    $this->set('parentView, $parentView);
  }
}

// cell template
$parentView->Html->scriptBlock('alert("hi")', ['block' => 'blockName']);

// Rendering in the parent view
<?= $this->cell('SharedState::share', [$this]) ?>
```

I'm not convinced that  we should break the encapsulation/isolation when there are existing solutions.

**Comment by inoas** [comment]
But wouldn't passing the view into the cell open a can of worms of possibilities to break encapsulation/isolation? Especially as a logical child is mutating the state of its parent ($this).
vs... just allowing cells to do very specific things on opt in basis, e.g. to push things into css, cssBottom, script and scriptBottom?

**Comment by markstory** [comment]
@ionas Sure, but that's what is being asked for here. Cells are isolated on purpose, but this entire thread is about opening that bubble so that cells can mutate state of the parent. Sharing blocks requires sharing view context, which means sharing the whole context.

**Comment by saeideng** [comment]
What you thing about somting like
@this->cell('Menu')->extractBlocks()->render():
When we used extrackBlocks() or withBlocks() cell's and main blocks can be merge

**Comment by saeideng** [comment]
Or on $this
$this->withBlocks->cell('....')

**Comment by saeideng** [comment]
Above cods is just an idea
I am not sure thats can be work

**Comment by thinkingmedia** [comment]
This needs to be fixed without introducing special new methods that a developer needs to know about. Otherwise, it will be impossible for plugin developers to use the script block from inside their plugin's cells.

**Comment by thinkingmedia** [comment]
We could add `Blocks` to the `ViewBuilder` class, and then update `CellTrail` to inject blocks from the current view. 

I think this would be consistent with how cells are created already. The ViewBuilder is injected with the current theme, the current helpers and plugin name.

ViewBuilder would also need to pass `Blocks` to the View's constructor, but there is already a `$_passedVars` map in View that handles this.

Now, can anyone think of a reason why this would break anything?

**Comment by markstory** [comment]
Well you'd be sharing view state that wasn't shared before :smile:

I am have no doubts that we can come up with a variety of ways to make cells share state with the parent context. I still come back to why this is necessary. Are there use cases beyond helpers being used to emit javascript code?

**Comment by inoas** [comment]
That's the reason I was suggesting to add a `viewScope` attribute that could be set to 'global' - it it wasn't then cells would be entirely contained.

1. requiring to eager or lazy load javascript, css
2. pushing http2 prefetch javascript ressourches through the response head 
3. putting html snipplets at the top or bottom of the document (as anchors, say bootstrap modal, while I think little of it)

I'd probably disregard 3. as wontfix and make sure there is support for 1. (for instance through a scope attribute that has to be explicitly set) and 2. (here cell's 'actions' have to have access to the http response head).

E.g. I'd try to limit what cells can do to injecting css/js into the head / below the fold, then also make sure cells can utilize http2 prefetch to supply additional css/js ressources cells could need

**Comment by othercorey** [comment]
@saeideng Do you still expect this to happen?

**Comment by saeideng** [comment]
yes , and milestone correctly  is `Future`

**Comment by othercorey** [comment]
> yes , and milestone correctly is `Future`

What solution do you have in mind?

**Comment by peterbaumert** [comment]
Still does not seem to work for me.

I have this in my cells display.php:

```
$SVG = <<<SVG
<symbol viewBox="0 0 16 16">
 removed
</symbol>
SVG;

$this->append('icons', $SVG);
```

and this in my layout:

```
 <?php echo $this->fetch('icons'); ?>
```

Yet its just empty.

**Comment by LordSimal** [comment]
@peterbaumert as already mentioned previously this is intentional because Cells are not rendered with same "context" as your "normal" template files. See the yellow box on https://book.cakephp.org/4/en/views/cells.html#implementing-the-cell

Therefore your blocks inside your cell template can't be fetched in your "normal" template.

"Fixing" this would introduce a behavior change which would most likely break many other apps since they expect exactly this encapsulated behavior.

But you can definitely copy/paste that block code from inside your cell to the place where you call that cell.
So something like
```
<?php
echo $this->cell('MyCell');

$SVG = <<<SVG
<symbol viewBox="0 0 16 16">
 removed
</symbol>
SVG;

$this->append('icons', $SVG);
```
or if you need that combo more often then I would recommend building an [element](https://book.cakephp.org/4/en/views.html#elements) which you then call inside your template wherever you like

**Comment by peterbaumert** [comment]
Thanks for the fast reply. Then i understood it wrong. I thought thats what was "fixed" in 4.4. thanks again for the explanation
